catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/SourcePoller.java

```diff
@@ -13,110 +13,605 @@
+          pollingExecutorService,
+          EXECUTOR_SERVICE_TERMINATION_TIMEOUT,
+          EXECUTOR_SERVICE_TERMINATION_TIMEOUT_TIME_UNIT);
+    }
+
+    MoreExecutors.shutdownAndAwaitTermination(
+        sourceAvailabilityChecksThreadPool,
+        EXECUTOR_SERVICE_TERMINATION_TIMEOUT,
+        EXECUTOR_SERVICE_TERMINATION_TIMEOUT_TIME_UNIT);
+  }
 
   /**
-   * Constructor to schedule the SourcePollerRunner to execute immediately and at a fixed interval,
-   * currently set at every 1 minute. This constructor is invoked by the CatalogFramework's
-   * blueprint.
+   * This method is an non-blocking alternative to {@link Source#isAvailable()} which provides more
+   * more details about the availability than just available or unavailable. The last availability
+   * check is guaranteed to be started when the source is bound with {@link #bind(Source)} or at the
+   * start of the lat poll interval, configured by {@link #setPollIntervalMinutes(long)}.
    *
-   * @param incomingRunner the SourcePollerRunner to use for polling
+   * @return a {@link SourceAvailability} for the availability of the {@code source}, or {@link
+   *     Optional#empty()} if the availability of the {@code source} has not yet been checked
+   * @throws IllegalArgumentException if the {@code source} is {@code null}
    */
-  public SourcePoller(SourcePollerRunner incomingRunner) {
+  public Optional<SourceAvailability> getSourceAvailability(final Source source) {
+    notNull(source);
 
-    this.runner = incomingRunner;
+    LOGGER.trace("Getting status of source id={}", source.getId());
+    return Optional.ofNullable(cache.getIfPresent(new SourceKey(source)))
+        .flatMap(CachedSourceAvailability::getSourceAvailability);
+  }
 
-    scheduler =
+  /**
+   * Starts a process to periodically check the availability of each of the {@link Source}s
+   * configured by {@link #setConnectedSources(List)}, {@link #setFederatedSources(List)}, {@link
+   * #setCatalogProviders(List)}, and {@link #setCatalogStores(List)}.
+   *
+   * <p>At the start of each interval, a process to check the availability of a {@link Source} will
+   * not be started if the {@link Source} was bound or modified since the last poll and the
+   * availability check started by that action has not yet completed.
+   *
+   * @param pollIntervalMinutes the interval (in minutes) at which to recheck the availability of
+   *     the {@link Source}s
+   * @throws IllegalArgumentException if {@code pollIntervalMinutes} is less than {@code 1}
+   */
+  public void setPollIntervalMinutes(final long pollIntervalMinutes) {
+    final long minimumPollIntervalMinutes = 1;
+    if (pollIntervalMinutes < minimumPollIntervalMinutes) {
+      throw new IllegalArgumentException(
+          "pollIntervalMinutes argument may not be less than " + minimumPollIntervalMinutes);
+    }
+
+    if (pollingExecutorService != null) {
+      MoreExecutors.shutdownAndAwaitTermination(
```

Comment:
If we pass in the scheduled executor and then just hang on to the future we can stop the job but not the executor.